### PR TITLE
Add allow_browser to set minimum versions for your application

### PR DESCRIPTION
**Pull request by dhh**
Specify the browser versions that will be allowed to access all actions (or some, as limited by <tt>only:</tt> or <tt>except:</tt>).

Only browsers matched in the hash or named set passed to <tt>versions:</tt> will be blocked if they're below the versions specified.

This means that all other browsers, as well as agents that aren't reporting a user-agent header, will be allowed access.

A browser that's blocked will by default be served the file in public/426.html with a HTTP status code of "426 Upgrade Required".

In addition to specifically named browser versions, you can also pass <tt>:modern</tt> as the set to restrict support to browsers natively supporting webp images, web push, badges, import maps, CSS nesting, and CSS :has. This includes Safari 17.2+, Chrome 119+, Firefox 121+, Opera 104+.

You can use https://caniuse.com to check for browser versions supporting the features you use.

You can use <tt>ActiveSupport::Notifications</tt> to subscribe to events of browsers being blocked using the <tt>browser_block.action_controller</tt> event name.

Examples:

```ruby
class ApplicationController < ActionController::Base
  # Allow only browsers natively supporting webp images, web push, badges, import maps, CSS nesting + :has 
  allow_browser versions: :modern
end

class ApplicationController < ActionController::Base
  # All versions of Chrome and Opera will be allowed, but no versions of "internet explorer" (ie). Safari needs to be 16.4+ and Firefox 121+.
  allow_browser versions: { safari: 16.4, firefox: 121, ie: false }
end

class MessagesController < ApplicationController
  # In addition to the browsers blocked by ApplicationController, also block Opera below 104 and Chrome below 119 for the show action.
  allow_browser versions: { opera: 104, chrome: 119 }, only: :show
end
```

### CODE REVIEW

**Comment by dhh** [comment]
Again, flaky, unrelated test:

Screenshot 2023-12-31 at 18 47 39

File: railties/lib/rails/generators/rails/app/templates/public/426.html [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  }
+  </style>
+</head>
+
+<body class="rails-default-error-page">
+  <!-- This file lives in public/426.html -->
+  <div class="dialog">
+    <div>
+      <h1>Your browser is not supported.</h1>
+      <p>Please upgrading your browser to continue.</p>
```
Comment (codergeek121) [inline railties/lib/rails/generators/rails/app/templates/public/426.html:62]:
```suggestion
      <p>Please upgrade your browser to continue.</p>
```

Small suggestion :)

I like the API/implementation, it's still flexible to support a gradual deployment.

**Comment by zzak** [comment]
This should have a changelog, if anyone wants to add it :pray:

**Comment by ghiculescu** [comment]
FYI https://twitter.com/aviflombaum/status/1746985145884619205

seems wrong

**Comment by dhh** [comment]
How so? He’s using a custom user agent string that says he’s using an old browser. Looks like this is working as designed.

**Comment by dhh** [comment]
It's not about the phone, it's about the fact that the user agent says that phone is running iOS 14 or something. An iPhone 12 Pro can be upgraded to the latest iOS no problem.

**Comment by dhh** [comment]
iOS 16.4 is supported by: iPhone 14, iPhone 14 Plus, iPhone 14 Pro, iPhone 14 Pro Max, iPhone 13, iPhone 13 mini, iPhone 13 Pro, iPhone 13 Pro Max, iPhone 12, iPhone 12 mini, iPhone 12 Pro, iPhone 12 Pro Max, iPhone 11, iPhone 11 Pro, iPhone 11 Pro Max, iPhone XS, iPhone XS Max, iPhone XR, iPhone X, iPhone 8, iPhone 8 Plus, iPhone SE (3rd generation), and iPhone SE (2nd generation). So that's phones 5-6 years old.

Anyway, this is a generated default. People can change it if they want something more lax.

**Comment by ghiculescu** [comment]
Using a Safari on iOS 16 user agent from https://whatmyuseragent.com/platforms/ios/ios/16

```ruby
3.2.2 :008 > UserAgent.parse("Mozilla/5.0 (iPhone; CPU iPhone OS 16_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.0 Mobile/15E148 Safari/604.1").version
 => #<UserAgent::Version 16.0>
```

So iOS 16 is blocked by default (this PR requires Safari >= 17.2). Would you consider a PR to change the default?

**Comment by dhh** [comment]
Ah yes, I see that we're on with 17.2. That'll give you access to all the latest stuff including css :has. I think we should be skating to where the puck is going to. We're not five minutes from release, so would prefer to keep it where it is. Either way, it's an exposed option, so you can always change it.

**Pull request merged**